### ISSUE TITLE
Array subdocument validation does not work in certain conditions

### ISSUE THREAD

**Issue by juona** [Automattic/mongoose#7117]
Mongoose 5.2.8 + Mongo 4.0.1

This seems like an odd **bug** with differing outcomes in only slightly different scenarios. I have found some extremely weird behaviors, so this is going to be a long bug report, please bear with me.

First of all, here is my code (a simplified example I wrote to test this specifically):

```
import mongoose, { Schema } from "mongoose";

const TestArrayItemSchema = new Schema({
  isNameNotRequired: {
    type: Boolean,
    required: true,
    default: true
  },
  name: {
    type: String,
    required() {
      return !this.isNameNotRequired;
    }
  }
});

const TestSchema = new Schema({
  testArray: [TestArrayItemSchema]
});

TestSchema.statics = {
  createNew() {
    return new TestModel().save();
  },

  addArrayItem(id) {
    return this.findById(id)
      .then(instance => {
        instance.testArray.push({});
        return instance;
      })
      .then(instance => instance.save())
      .then(instance => instance.testArray[instance.testArray.length - 1]);
  },

  updateArrayItem(instanceID, arrayItemID, details) {
    return this.findOne({
      _id: instanceID,
      "testArray._id": arrayItemID
    })
      .then(instance => {
        const arrayItem = instance.testArray.id(arrayItemID);
        Object.assign(arrayItem, details);
        return instance;
      })
      .then(instance => instance.save())
      .then(instance => instance.testArray.id(arrayItemID));
  }
};

const TestModel = mongoose.model("Test", TestSchema, "testing");

export default TestModel;
```

Quick summary. The schema is as follows:

```
{
	testArray: [{
		name: "abcd",
		isNameNotRequired: false
	}]
}
```

There are three static methods. I have used `createNew` to instantiate an empty TestModel object for further testing and didn't use it again. `addArrayItem` adds a new default TestArrayItem to testArray. `updateArrayItem` fetches the whole TestModel and updates the indicated TestArrayItem with the provided values before calling `save()`. I do this to make sure that the validators run correctly (in contract to update validators).

Expected behavior: whenever I call `updateArrayItem` the required validator should be run on the `name` field and work as expected, regardless of:

- how the two fields are named; e.g. if I change `name` to `myField` and `isNameNotRequired` to `anotherField`, should still work;
- how many items I have in the array;
- number of times I call this method;
- if I have multiple array items and call this method on them one-by-one - regardless of the order of the calls.

### Scenario 1

Make sure that for the chosen TestModel `testArray` is **empty** (as you shall see later, the outcome would differ for non-empty array). Add a new ArrayItem by calling `addArrayItem`.

Call `updateArrayItem` with { isNameNotRequired: false, name: null }.

Expected behavior - error.
Current behavior - saves to Mongo => FAIL.

Call `updateArrayItem` again without changing anything - parameters etc.

Expected behavior - error.
Current behavior - error => OK. All subsequent calls produce an error.

### Scenario 2

Make sure that for the chosen TestModel `testArray` is **empty**. Add a new ArrayItem by calling `addArrayItem`.

Call `updateArrayItem` with { isNameNotRequired: false }, omitting the name property altogether.

Expected behavior - error.
Current behavior - error => OK.

### Scenario 3

This is rather weird, please bear with me.

1. Like before, make sure that `testArray` is empty.
2. Add **at least two** ArrayItems by using `addArrayItem` at least twice.
3. Chose whichever item in the array to update and call `updateArrayItem` with { isNameNotRequired: false, name: null } for that item.

Expected - error.
Actual - saves to Mongo successfully => FAIL.

4. Try the operation again.

Expected - error.
Actual - error => FAIL! All subsequent calls work fine. At this point, if I remove the name property from the update request, it still works. Server + Mongo restart does not help.

5. Try to run a bad update on any other item in the array.

Expected - error.
Actual - error => OK. All other items in the array work correctly regardless of the order in which they are chosen. Only the one that is chose for the first update operation remains broken.

### Scenario 4

If **scenario 3** is repeated but `updateArrayItem` is called with { isNameNotRequired: false }, i.e. without the name property, it works as expected - errors are thrown in all cases.

### Other information

1. You may not be able to reproduce this right away. If you copy my code and try it out, it might work. To make it fail, please try one or both of the following:
- Rename the `name` field to something else, like `name1`. This also had me very confused, because it seemed that my initial code worked, but any modifications to it didn't. It still seems that this is true, but it would be an overkill to add that to this issue.
- If that does not help, delete the created collection, and restart the servers, which are running (like Mongo or your local dev - I was testing the code using a GraphQL API since it seemed a lot faster than writing tests; possibly only deleting the collection itself is enough).

2. I spent about a day inspecting Mongoose for and here are my findings:
- Whenever the code does not work as expected, the required method for the `name` property is not called **at all**.
- It may be that the reason is that there is a check for `$cachedRequired`, which determines whether or not a particular validator should be executed. This check is present in order to avoid calling the validator twice in the same validate call (schematype.js:652).
- The implementation here seems correct, but the problem is that when the document is retrieved by the `findOne` method (in my code), it is already returned with `$cachedRequired` set, which is what ultimately bypasses the required validator method.
- This is not always the case, however, e.g. in **scenario 3**, so I may be completely wrong.

It took me a while to wrap my head around these issues. There might be more similar scenarios.

**Comment by vkarpov15** [comment]
Thanks for the detailed bug report. Have you tried upgrading to mongoose 5.2.18? I believe we made some fixes related to $cachedRequired since 5.2.8.

**Comment by vkarpov15** [comment]
No worries, thanks for looking into it.
